source/documentNavigation/sentenceHelper.py

```diff
@@ -0,0 +1,40 @@
+# -*- coding: UTF-8 -*-
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2022 NV Access Limited, Rob Meredith
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+import re
+from typing import Optional
+
+WESTERN_TERMINATORS_PATTERN: re.Pattern = re.compile("[.?!](?: |\r|\n)+")
+WESTERN_TERMINATORS_SINGLE_PATTERN: re.Pattern = re.compile("[.?!]")
+FULL_WIDTH_TERMINATORS_PATTERN: re.Pattern = re.compile("[。！？](?:\r|\n)*")
+
+
+def _findEndOfSentence(sentence: str, offset: int) -> Optional[int]:
+	"""
+	Returns an index which points one character past the end of the current sentence,
+	or if the sentence is at the end of the string, returns the length of the string.
+	Returns None if sentence end not found.
+	In the case of western sentence terminators,
+	conbinations of space, \r, or \n at the end of the sentence are included.
+	In the case of full width sentence terminators,
+	conbinations of \r or \n at the end of the sentence are included.
+	@param sentence: string containing at least current sentence
+	@param offset: int indicating where to start searching for the end of the sentence
+	"""
+	res = None
+	if not len(sentence):
+		return res  # do nothing if empty string
+	m = re.search(FULL_WIDTH_TERMINATORS_PATTERN, sentence[offset:])
```

Comment:
Currently, in the case of a long text consisting in many western sentences followed by a Chinese sentence at the very end, the western ends of sentence will be skipped and the Chinese end of sentence will directly match.

Instead, we expect the first end of sentence to match, that is, here, the first western end of sentence.

A unit test consisting in a western sentence followed by a Chinese sentence would be interesting in this case.

To address this issue, you should create a unique regexp combining full width and western end of sentences by or-ing all the possible ends of sentence. Then make a unique match with this regexp.

Reply:
> @CyrilleB79 While what you state is correct, I'm not sure strings containing both western and full widh terminators is a likely scenario. I think its fine as is. I wonder what @seanbudd or @feerrenrut think?

I agree that mixing sentences of different languages is probably not so common (except in situations like language teaching material).

I had just made this suggestion because changing the code to support this is quite simple:
* or-ing the two regexp strings before compiling the resulting string to a re.Pattern
* using this pattern to make only one re.search instead of two in the function

This would also allow to add more easily other patterns if needed in the future. E.g. Arabic punctuation could be added without modifying the function if this community provide more information on the rules of this language.

Reply:
@CyrilleB79 OK, I took your advice and or-ed the expressions before compiling, and added a unit test for the case of mixed sentence types.